### PR TITLE
Specify additional requirements for LLVM module

### PR DESCRIPTION
**Pull request by bader**

### CODE REVIEW

File: docs/SPIRVRepresentationInLLVM.rst [hunk 1]
```diff
@@ -213,6 +229,53 @@ following format:
...
+| OpSourceExtension  | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.SourceExtension = !{!0, !1}                   |
+|                    |    !0 = !{!"cl_khr_fp16"}                               |
+|                    |    !1 = !{!"cl_khr_gl_sharing"}                         |
++--------------------+---------------------------------------------------------+
+| OpExtension        | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.Extension = !{!0}                             |
+|                    |    !0 = !{!"cl_khr_fp16"}                               |
```
Comment (AlexeySotkin) [inline docs/SPIRVRepresentationInLLVM.rst:253]: `cl_khr_fp16` is a _source_ extension. I suggest using a SPIR-V extension, for example:
```suggestion
|                    |    !0 = !{!"SPV_KHR_expect_assume"}                               |
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 2]
```diff
@@ -213,6 +229,53 @@ following format:
...
++--------------------+---------------------------------------------------------+
+| OpCapability       | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.Capability = !{!0}                            |
+|                    |    !0 = !{i32 10} ; Float64 - program uses doubles      |
++--------------------+---------------------------------------------------------+
+| OpExecutionMode    | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.ExecutionMode = !{!0}                         |
+|                    |    !6 = !{void ()* @worker, i32 30, i32 262149}         |
```
Comment (AlexeySotkin) [inline docs/SPIRVRepresentationInLLVM.rst:263]:
```suggestion
|                    |    !0 = !{void ()* @worker, i32 30, i32 262149}         |
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 3]
```diff
@@ -213,6 +229,53 @@ following format:
...
+|                    |    ; Set execution mode with id 30 (VecTypeHint) and    |
+|                    |    ; literal `262149` operand.                          |
++--------------------+---------------------------------------------------------+
+| OpCapability       | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.Capability = !{!0}                            |
+|                    |    !0 = !{i32 10} ; Float64 - program uses doubles      |
++--------------------+---------------------------------------------------------+
```
Comment (AlexeySotkin) [inline docs/SPIRVRepresentationInLLVM.rst:271]: This is a duplicate row in the table. Should we remove it?

File: docs/SPIRVRepresentationInLLVM.rst [hunk 4]
```diff
@@ -213,6 +229,53 @@ following format:
...
+|                    |                                                         |
+|                    |    !spirv.ExecutionMode = !{!0}                         |
+|                    |    !6 = !{void ()* @worker, i32 30, i32 262149}         |
+|                    |    ; Set execution mode with id 30 (VecTypeHint) and    |
+|                    |    ; literal `262149` operand.                          |
++--------------------+---------------------------------------------------------+
+| OpCapability       | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.Capability = !{!0}                            |
+|                    |    !0 = !{i32 10} ; Float64 - program uses doubles      |
```
Comment (bader) [inline docs/SPIRVRepresentationInLLVM.rst:270]:
```suggestion
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 5]
```diff
@@ -74,6 +75,21 @@ mangled as __spirv_{TypeName}, where {TypeName} is the name of the SPIR-V
...
 mangled as __spirv_Event.
 
+Address spaces
+--------------
+
+Following
+`SPIR-V storage classes <https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Storage_Class>`_
+are naturally represented as LLVM IR address spaces with following mapping:
+
+SPIR-V storage class    LLVM IR address space
```
Comment (MrSidims) [inline docs/SPIRVRepresentationInLLVM.rst:85]: What about storage classes added via extensions? For example Device/Host storages from SPV_INTEL_usm_storage_classes, which are mapped on addrspace(5) and addrspace(6) appropriately?

File: docs/SPIRVRepresentationInLLVM.rst [hunk 6]
```diff
@@ -239,3 +297,55 @@ For example:
...
+
+For example:
+
+- !kernel_arg_addr_space !1
+- !kernel_arg_access_qual !2
+- !kernel_arg_type !3
+- !kernel_arg_base_type !4
+- !kernel_arg_type_qual !5
+
+**TODO**: Are all these metadata optional? Access qualifiers are translated for
```
Comment (MrSidims) [inline docs/SPIRVRepresentationInLLVM.rst:344]: Summoning @StuartDBrady for a discussion.

File: docs/SPIRVRepresentationInLLVM.rst (hunk 5 above)
Comment (MrSidims) [reply docs/SPIRVRepresentationInLLVM.rst:85]: By 'mapped' I mean they are mapped on global_device and global_host address spaced in OpenCL (llvm.org) and in OpenCL/DPCPP (intel/llvm).

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:85]: I'm not sure I understand your question.
Do you suggest adding "the mapping" to this document or just want to learn something about these address spaces?
Could you specify the question, please?

Comment (MrSidims) [reply docs/SPIRVRepresentationInLLVM.rst:85]: If I understand the document's purpose correctly, it is about 'what LLVM module can (and probably must) be correctly translated to SPIR-V. Additional requirements section introduces a restriction on what LLVM address spaces are valid. I'm saying, that for some (read: USM) extensions it's valid to have other address spaces other than 0-4 in spir LLVM IR.

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:85]: It's not clear what exactly do you imply with this comment, but I added an example with `SPV_INTEL_usm_storage_classes` extension to the document. Let me know if you expected something else.

**Review by MrSidims — APPROVED** [review]
LGTM

**Review by AlexeySachkov — APPROVED** [review]

File: docs/SPIRVRepresentationInLLVM.rst [hunk 7]
```diff
@@ -74,6 +75,24 @@ mangled as __spirv_{TypeName}, where {TypeName} is the name of the SPIR-V
 type with "OpType" removed, e.g., OpTypeEvent is mapped to spirv.Event and
 mangled as __spirv_Event.
 
+Address spaces
+--------------
+
+Following
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:81]:
```suggestion
The following
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 8]
```diff
@@ -74,6 +75,24 @@ mangled as __spirv_{TypeName}, where {TypeName} is the name of the SPIR-V
...
+`SPIR-V storage classes <https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Storage_Class>`_
+are naturally represented as LLVM IR address spaces with following mapping:
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:83]:
```suggestion
are naturally represented as LLVM IR address spaces with the following mapping:
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 9]
```diff
@@ -239,3 +300,55 @@ For example:
...
+
+- ``spir`` -> Physical32
+- ``spir64`` -> Physical64
+
+Calling convention
+------------------
+
+OpEntryPoint information is represented in LLVM IR in calling convention.
+A function with `spir_kernel` calling convention will be translated as an entry
+point to SPIR-V module.
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:332]:
```suggestion
point of the SPIR-V module.
```

File: docs/SPIRVRepresentationInLLVM.rst (hunk 6 above)
Comment (svenvh) [reply docs/SPIRVRepresentationInLLVM.rst:347]: > Are all these metadata optional?

I think so.  Clang started generating this metadata as of commit `fb36ede52ebb ("Generate metadata to implement the -cl-kernel-arg-info option. OpenCL 1.2 spec. 5.7.3.", 2013-03-24)`, with the purpose of supporting `clGetKernelArgInfo`.  The OpenCL API spec says that 

> Kernel argument information is only available if the program object associated with kernel is created with clCreateProgramWithSource and the program executable was built with the -cl-kernel-arg-info option specified in options argument to clBuildProgram or clCompileProgram.

So an implementation should not need to support kernel argument info in the SPIR-V case.

File: docs/SPIRVRepresentationInLLVM.rst [hunk 10]
```diff
@@ -213,6 +232,48 @@ following format:
   !<InstructionMetadata1> = !{<Operand1>, <Operand2>, ..}
   !<InstructionMetadata2> = !{<Operand1>, <Operand2>, ..}
 
++--------------------+---------------------------------------------------------+
```
Comment (AlexeySotkin) [inline docs/SPIRVRepresentationInLLVM.rst:235]: This table is not rendered in the rich diff

File: docs/SPIRVRepresentationInLLVM.rst (hunk 6 above)
Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst]: Changed TODO to NOTE.

File: docs/SPIRVRepresentationInLLVM.rst (hunk 10 above)
Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:235]: Thanks for pointing this. Should be fixed by cfc7c8a8dc7ec4f45a530840c4a9472ea6a0af5f.

**Review by AlexeySotkin — APPROVED** [review]

File: docs/SPIRVRepresentationInLLVM.rst [hunk 11]
```diff
@@ -74,6 +75,24 @@ mangled as __spirv_{TypeName}, where {TypeName} is the name of the SPIR-V
...
+SPIR-V storage class    LLVM IR address space
+Function                No address space or addrspace(0)
+CrossWorkgroup          addrspace(1)
+UniformConstant         addrspace(2)
+Workgroup               addrspace(3)
+Generic                 addrspace(4)
+
+SPIR-V extensions are allowed to add new storage classes. For example,
+SPV_INTEL_usm_storage_classes extension adds DeviceOnlyINTEL and HostOnlyINTEL
+storage classes which are mapped to addrspace(5) and addrspace(6) appropriately.
```
Comment (StuartDBrady) [inline docs/SPIRVRepresentationInLLVM.rst:94]: Do you mean "respectively" instead of "appropriately"?

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:94]:
```suggestion
storage classes which are mapped to addrspace(5) and addrspace(6) respectively.
```

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst]: Thanks.
Fixed.

File: docs/SPIRVRepresentationInLLVM.rst [hunk 12]
```diff
@@ -213,6 +232,48 @@ following format:
   !<InstructionMetadata1> = !{<Operand1>, <Operand2>, ..}
   !<InstructionMetadata2> = !{<Operand1>, <Operand2>, ..}
 
++--------------------+---------------------------------------------------------+
+| SPIR-V instruction | LLVM IR                                                 |
++====================+=========================================================+
+| OpSource           | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.Source = !{!0}                                |
+|                    |    !0 = !{i32 3, i32 102000, !1}                        |
```
Comment (StuartDBrady) [inline docs/SPIRVRepresentationInLLVM.rst:241]: Could I ask where the numbering of OpenCL C versions for SPIR-V is specified?

I can see that it's defined within the SPIR-V/LLVM translator in lib/SPIRV/OCLUtil.h as follows:
```
const unsigned CL12 = 102000;
const unsigned CL20 = 200000;
const unsigned CL21 = 201000;
const unsigned CL30 = 300000;
```
However, that doesn't tell me why.

I find this confusing, as I would expect -cl-std=CL1.2 to be 120 or 10200, and not 102000.

The SPIR-V spec gives `OpSource GLSL 450` as an example, and the tests in SPIR-V Tools include e.g. `OpSource OpenCL_C 120`, `OpSource OpenCL_C 12` and ``OpSource OpenCL_C 1202000`.

Note that the SPIR-V file header encodes the SPIR-V version as bytes, e.g. 0x00010300 for SPIR-V 1.3, but this is the _SPIR-V_ version, and it is encoded with a _byte per component_, and not as a pair of decimal digits per component, and there isn't the extra trailing zero.

My concern here is whether we want to standardize what seems to be a bug.

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:241]: > Could I ask where the numbering of OpenCL C versions for SPIR-V is specified?

https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_Env.html#_source_language_encoding

I think I got this example from one of tests, so there might be bugs. Probably the format provided there is quite outdated.
If I understand OpenCL environment specification correctly, the 32-bit literal consists of 4 bytes: 0 | Major Number | Minor Number | Revision Number (optional). 
OpenCL 1.2 -> hex (0x00010200) -> dec 66048.

```suggestion
|                    |    !0 = !{i32 3, i32 66048, !1}                        |
```
Right?

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:241]: @StuartDBrady, are you okay to commit the change suggested in my previous comment?

Comment (StuartDBrady) [reply docs/SPIRVRepresentationInLLVM.rst:241]: The suggestion looks fine to me.

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst]: Thanks. I've applied the change.

**Comment by bader** [comment]
@svenvh, @StuartDBrady, please, let me know if it looks good to you.

File: docs/SPIRVRepresentationInLLVM.rst [hunk 13]
```diff
@@ -74,6 +75,24 @@ mangled as __spirv_{TypeName}, where {TypeName} is the name of the SPIR-V
...
 
+Address spaces
+--------------
+
+The following
+`SPIR-V storage classes <https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Storage_Class>`_
+are naturally represented as LLVM IR address spaces with the following mapping:
+
+SPIR-V storage class    LLVM IR address space
+Function                No address space or addrspace(0)
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:86]: This doesn't render as a table currently.

It's also nicer to render the keywords as inline literals using ` `` `

File: docs/SPIRVRepresentationInLLVM.rst (hunk 10 above)
Comment (svenvh) [reply docs/SPIRVRepresentationInLLVM.rst:235]: This table is still not rendered when clicking "View file".

File: docs/SPIRVRepresentationInLLVM.rst [hunk 14]
```diff
@@ -239,3 +300,55 @@ For example:
...
+SPIR-V instruction.
+
+- ``spir`` -> Physical32
+- ``spir64`` -> Physical64
+
+Calling convention
+------------------
+
+OpEntryPoint information is represented in LLVM IR in calling convention.
+A function with `spir_kernel` calling convention will be translated as an entry
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:331]:
```suggestion
A function with ``spir_kernel`` calling convention will be translated as an entry
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 15]
```diff
@@ -239,3 +300,55 @@ For example:
...
+point of the SPIR-V module.
+
+Function metadata
+-----------------
+
+Some kernel parameter information is stored in LLVM IR as a function metadata.
+
+For example:
+
+- !kernel_arg_addr_space !1
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:341]: This could be a code-block I guess.

File: docs/SPIRVRepresentationInLLVM.rst [hunk 16]
```diff
@@ -239,3 +300,55 @@ For example:
...
+`OpMemoryModel <https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_mode_setting_a_mode_setting_instructions>`_
+SPIR-V instruction.
+
+- ``spir`` -> Physical32
+- ``spir64`` -> Physical64
+
+Calling convention
+------------------
+
+OpEntryPoint information is represented in LLVM IR in calling convention.
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:330]:
```suggestion
``OpEntryPoint`` information is represented in LLVM IR in calling convention.
```

File: docs/SPIRVRepresentationInLLVM.rst [hunk 17]
```diff
@@ -239,3 +300,55 @@ For example:
...
+
+Target triple and datalayout string
+-----------------------------------
+
+Target triple architecture must be ``spir`` (32-bit architecture) or ``spir64``
+(64-bit architecture) and ``datalayout`` string must be aligned with OpenCL
+environment specification requirements for data type sizes and alignments (e.g.
+3-element vector must have 4-element vector alignment). For example:
+
+.. code-block:: llvm
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:314]: This example isn't rendered.

File: docs/SPIRVRepresentationInLLVM.rst [hunk 18]
```diff
@@ -213,6 +232,48 @@ following format:
...
 
++--------------------+---------------------------------------------------------+
+| SPIR-V instruction | LLVM IR                                                 |
++====================+=========================================================+
+| OpSource           | .. code-block:: llvm                                    |
+|                    |                                                         |
+|                    |    !spirv.Source = !{!0}                                |
+|                    |    !0 = !{i32 3, i32 66048, !1}                        |
+|                    |    ; 3 - OpenCL_C                                       |
+|                    |    ; 102000 - OpenCL version 1.2                        |
```
Comment (svenvh) [inline docs/SPIRVRepresentationInLLVM.rst:243]:
```suggestion
|                    |    ; 66048 = 0x10200 - OpenCL version 1.2               |
```

**Review by svenvh — COMMENTED** [review]
Mostly formatting/rendering issues left I think.  I'm not sure what you used to view the rendered version, but in the GitHub renderer quite a few of your additions don't show up properly (especially tables).

File: docs/SPIRVRepresentationInLLVM.rst (hunk 13 above)
Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:86]: Which words are keywords? `addrspace(x)`? Or SPIR-V storage classes as well (i.e. `Function`)?

File: docs/SPIRVRepresentationInLLVM.rst (hunk 10 above)
Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:238]: It was broken again by https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/93ac5b2d110304b5975fcdfe4286b908910e7c11, but https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/16df0913330b960bcb118648c63c06e7579bb400 fixes it.

File: docs/SPIRVRepresentationInLLVM.rst (hunk 17 above)
Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst:317]: Fixed with https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/16df0913330b960bcb118648c63c06e7579bb400.

File: docs/SPIRVRepresentationInLLVM.rst (hunk 15 above)
Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst]: Applied in https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/16df0913330b960bcb118648c63c06e7579bb400.

File: docs/SPIRVRepresentationInLLVM.rst (hunk 13 above)
Comment (svenvh) [reply docs/SPIRVRepresentationInLLVM.rst]: I'd suggest both ``addrspace(x)`` and SPIR-V storage classes.

Comment (bader) [reply docs/SPIRVRepresentationInLLVM.rst]: Done.

**Review by svenvh — APPROVED** [review]
LGTM

**Review by AlexeySotkin — APPROVED** [review]

**Pull request merged**